src/main/java/io/vertx/core/eventbus/impl/CodecManager.java

```diff
@@ -46,103 +46,111 @@
+  private final ConcurrentMap<Class, MessageCodec> codecsByClass = new ConcurrentHashMap<>();
 
   public CodecManager() {
-    this.systemCodecs = codecs(NULL_MESSAGE_CODEC, PING_MESSAGE_CODEC, STRING_MESSAGE_CODEC, BUFFER_MESSAGE_CODEC, JSON_OBJECT_MESSAGE_CODEC, JSON_ARRAY_MESSAGE_CODEC,
+    this.systemCodecs = new MessageCodec[] { NULL_MESSAGE_CODEC, PING_MESSAGE_CODEC, STRING_MESSAGE_CODEC, BUFFER_MESSAGE_CODEC, JSON_OBJECT_MESSAGE_CODEC, JSON_ARRAY_MESSAGE_CODEC,
       BYTE_ARRAY_MESSAGE_CODEC, INT_MESSAGE_CODEC, LONG_MESSAGE_CODEC, FLOAT_MESSAGE_CODEC, DOUBLE_MESSAGE_CODEC,
-      BOOLEAN_MESSAGE_CODEC, SHORT_MESSAGE_CODEC, CHAR_MESSAGE_CODEC, BYTE_MESSAGE_CODEC, REPLY_EXCEPTION_MESSAGE_CODEC);
+      BOOLEAN_MESSAGE_CODEC, SHORT_MESSAGE_CODEC, CHAR_MESSAGE_CODEC, BYTE_MESSAGE_CODEC, REPLY_EXCEPTION_MESSAGE_CODEC };
   }
 
   public MessageCodec lookupCodec(Object body, String codecName) {
-    MessageCodec codec;
     if (codecName != null) {
-      codec = userCodecMap.get(codecName);
+      MessageCodec codec;
+      codec = codecsByName.get(codecName);
       if (codec == null) {
         throw new IllegalArgumentException("No message codec for name: " + codecName);
       }
-    } else if (body == null) {
-      codec = NULL_MESSAGE_CODEC;
-    } else if (body instanceof String) {
-      codec = STRING_MESSAGE_CODEC;
-    } else if (body instanceof Buffer) {
-      codec = BUFFER_MESSAGE_CODEC;
-    } else if (body instanceof JsonObject) {
-      codec = JSON_OBJECT_MESSAGE_CODEC;
-    } else if (body instanceof JsonArray) {
-      codec = JSON_ARRAY_MESSAGE_CODEC;
-    } else if (body instanceof byte[]) {
-      codec = BYTE_ARRAY_MESSAGE_CODEC;
-    } else if (body instanceof Integer) {
-      codec = INT_MESSAGE_CODEC;
-    } else if (body instanceof Long) {
-      codec = LONG_MESSAGE_CODEC;
-    } else if (body instanceof Float) {
-      codec = FLOAT_MESSAGE_CODEC;
-    } else if (body instanceof Double) {
-      codec = DOUBLE_MESSAGE_CODEC;
-    } else if (body instanceof Boolean) {
-      codec = BOOLEAN_MESSAGE_CODEC;
-    } else if (body instanceof Short) {
-      codec = SHORT_MESSAGE_CODEC;
-    } else if (body instanceof Character) {
-      codec = CHAR_MESSAGE_CODEC;
-    } else if (body instanceof Byte) {
-      codec = BYTE_MESSAGE_CODEC;
-    } else if (body instanceof ReplyException) {
-      codec = defaultCodecMap.get(body.getClass());
-      if (codec == null) {
-        codec = REPLY_EXCEPTION_MESSAGE_CODEC;
-      }
+      return codec;
     } else {
-      codec = defaultCodecMap.get(body.getClass());
-      if (codec == null) {
-        throw new IllegalArgumentException("No message codec for type: " + body.getClass());
+      if (body == null) {
+        return NULL_MESSAGE_CODEC;
+      } else {
+        Class<?> clazz = body.getClass();
+        if (clazz == String.class) {
+          return STRING_MESSAGE_CODEC;
+        } else if (Buffer.class.isAssignableFrom(clazz)) {
```

Comment:
Would it be better to check for BufferImpl here? This is an important difference to understand, because checking `isAssignableFrom` means that every implementation of `Buffer` is forced to used the build-in `BufferMessageCodec`. Different implementations of `Buffer` will not work. See also related [1].

[1] https://github.com/eclipse-vertx/vert.x/pull/4111/commits/acb42fafba589c42968bf2136395358195c2dfa8

Reply:
if we would not check Buffer, then any buffer implementation that is not a vertx heap buffer would be rejected

Reply:
Well yes, that would again speak for the `accept` method you suggested in your comment below, because due to the checks of the built-in codecs taking precedence over the custom codecs, your statements on the other hand means: every Buffer implementstion that is not a Vert.x heap buffer is forced to be converted into one, if it is sent via the event bus. I am fine with that, however that make custom implementations of `Buffer` very limited in their usefullness.